### PR TITLE
First PoC for dll debugging (don't merge)

### PR DESCRIPTION
**Pull request by xarkes**
So here's a quick test at debugging DLLs
I have many questions though:
* Is there any portable way to do a call ? So the code can be arch independant and each argument goes into its specific register, or get pushed to the stack or whatever depending on the arch.
* How can I detect if the file is a DLL ? (From within my current context of course -- Or maybe should I move this detection somewhere else ?)

LoadLib.exe is a simple binary which checks if the argv[1] is a file, and then calls LoadLibrary on it. I'll add it later in the repo.

Note that the mechanism is a bit different on Linux, there's no need to call a third-party binary to load a lib, so it's easier. The main problem is about doing a proper call on the function.

Any advice is welcome.

### CODE REVIEW

**Comment by skuater** [comment]
Hey you cant do this: 
r_core_cmd0 (core, "doo C:\\Windows\\System32\\user32.dll func args");

On windows u can use a rundll32.exe , this program allow to call dll funcs
rundll32.exe user32.dll, MessageBoxA "test"

Or a helper program todo this job.

**Comment by xarkes** [comment]
The problem with rundll is that radare2 always adds double quotes around the arguments.
I already tried and `rundll32.exe user32.dll,MessageBoxA "title" "body"` would actually  be translated into `rundll32.exe "user32.dll,MessageBoxA" "title" "body"` and for some reason, rundll does not like it and it just does not work.

**Comment by radare** [comment]
we shoudlnt add those quotes on windows.. or at least use the r_str_argv to parse that and pass the parameters properly

File: libr/io/p/io_debug.c [hunk 1]
```diff
@@ -96,8 +96,10 @@ static int setup_tokens() {
 }
 
 static int fork_and_ptraceme(RIO *io, int bits, const char *cmd) {
-	PROCESS_INFORMATION pi;
+	PROCESS_INFORMATION pi = { 0 };
```
Comment (mrexodia) [inline libr/io/p/io_debug.c]: I'm not sure this actually zeroes the whole struct. Doesn't it just set the first member to 0?

**Comment by radare** [comment]
yes it does

**Comment by radare** [comment]
at least in C and using clang and gcc, {0} initializes the whole structure even if it have more than one field or if its an array. i have checked this with valgrind and reading the disasm, and the whole struct is filled with memset

File: libr/io/p/io_debug.c [hunk 2]
```diff
@@ -96,8 +96,10 @@ static int setup_tokens() {
...
 	STARTUPINFOA si = { sizeof (si) };
```
Comment (mrexodia) [inline libr/io/p/io_debug.c]: In that case this makes no sense. Here you set all the members to "sizeof(si)" right?

Comment (xarkes) [inline libr/io/p/io_debug.c]: I agree it does not make sense, I don't even know who wrote this.

Comment (karliss) [inline libr/io/p/io_debug.c]: It sets first member to sizeof(si)  and rest of them 0.

**Comment by mrexodia** [comment]
So then the other one also only sets the first member to zero? Or is C
being inconsistent?

**Comment by xarkes** [comment]
`STARTUPINFOA si = { sizeof (si) };` "sets first member to sizeof(si) and rest of them 0"
`PROCESS_INFORMATION pi = { 0 };` sets first member to 0 and rest of them 0
I understood it like this.

**Comment by mrexodia** [comment]
Standard : If there are fewer initializers in a brace-enclosed list than
there are elements or members of an aggregate, or fewer characters in a
string literal used to initialize an array of known size than there are
elements in the array, the remainder of the aggregate shall be initialized
implicitly the same as objects that have static storage duration.

**Comment by mrexodia** [comment]
Yeah it's all dandy! 👌

File: libr/io/p/io_debug.c [hunk 3]
```diff
@@ -96,18 +96,20 @@ static int setup_tokens() {
...
 	if (!*cmd) {
 		return -1;
 	}
 	setup_tokens ();
-	char *_cmd = io->args ? r_str_appendf (strdup (cmd), " %s", io->args) :
-				strdup (cmd);
-	char **argv = r_str_argv (_cmd, NULL);
+	char **argv = r_str_argv (cmd, NULL);
+	/*char *_cmd = io->args ? r_str_appendf (strdup (cmd), " %s", io->args) : strdup (cmd);
 	// We need to build a command line with quoted argument and escaped quotes
```
Comment (xarkes) [inline libr/io/p/io_debug.c]: Ok but why ? I commented the whole block and it works the same (it does not parse any " etc.) but I don't understand why it was added in a first place.

**Comment by mrexodia** [comment]
It appears that `rundll32.exe` is rather unsuitable for this. See https://support.microsoft.com/en-us/help/164787/info-windows-rundll-and-rundll32-interface for documentation on how it's supposed to be used...

**Comment by XVilka** [comment]
@xarkes @mrexodia have you checked the approach of OllyDbg's LoadDLL? http://www.ollydbg.de/Loaddll.htm

**Comment by xarkes** [comment]
Okay so Olly's way is quite simple but I have a few questions though:
* Only x86 so they just push the arguments onto the stack and that's it. What about x86 64bits registers handling and maybe ARM later (iirc win10 runs on ARM too). Of course it's not much work to hardcode it but maybe is there a smarter way ?
* Their Dll Loader is written in assembly, so they can define symbols to their arguments pointer, so it is pre-allocated and the user can write it's argument (a char* for instance) there and Olly just gives the known pointer to the function. However it seems that with VS (I did not try with mingw) the binaries have no symbols and I'm still looking for a way to have those pointers available as flags into radare2. Example:
```Arg1       DB 1024 DUP (?)``` compiled with tasm32, the binary contains the Arg1 "symbol" (actually IDA describes it as an export, and I could not find it with radare2 lol)
```static const char Arg1[1024] = {0};``` compiled with VS, the binary does not contain the Arg1 symbol.

Tips appreciated :)

**Comment by mrexodia** [comment]
@xarkes I mentioned this before, but basically the only reason a DLL loader has to exist is that you need to call `LoadLibrary` on your DLL.

This code should do all you need:

```c++
#include <windows.h>

wchar_t szLibraryPath[512];

int main()
{
    memset(szLibraryPath, 0, sizeof(szLibraryPath));
    wchar_t szName[256] = L"";
    wsprintfW(szName, L"Local\\szLibraryName%X", (unsigned int)GetCurrentProcessId());
    HANDLE hMapFile = OpenFileMappingW(FILE_MAP_READ, false, szName);
    if(hMapFile)
    {
        const wchar_t* szLibraryPathMapping = (const wchar_t*)MapViewOfFile(hMapFile, FILE_MAP_READ, 0, 0, sizeof(szLibraryPath));
        if(szLibraryPathMapping)
        {
            lstrcpyW(szLibraryPath, szLibraryPathMapping);
            UnmapViewOfFile(szLibraryPathMapping);
        }
        CloseHandle(hMapFile);
    }
    if(szLibraryPath[0])
        return (LoadLibraryW(szLibraryPath) != NULL);
    return 0;
}
```

The code in r2 would create the file mapping and put a full path in it.

```c++
        PPROCESS_INFORMATION ReturnValue = (PPROCESS_INFORMATION)InitDebugW(szDebuggerName, szCommandLine, szCurrentFolder);
        wchar_t szName[256] = L"";
        swprintf(szName, 256, L"Local\\szLibraryName%X", (unsigned int)ReturnValue->dwProcessId);
        DebugDLLFileMapping = CreateFileMappingW(INVALID_HANDLE_VALUE, 0, PAGE_READWRITE, 0, 512 * sizeof(wchar_t), szName);
        if(DebugDLLFileMapping)
        {
            wchar_t* szLibraryPathMapping = (wchar_t*)MapViewOfFile(DebugDLLFileMapping, FILE_MAP_ALL_ACCESS, 0, 0, 512 * sizeof(wchar_t));
            if(szLibraryPathMapping)
            {
                wcscpy(szLibraryPathMapping, DebugDebuggingDLLFullFileName);
                UnmapViewOfFile(szLibraryPathMapping);
            }
        }
        ResumeThread(ReturnValue->hThread);
```

Obviously there are other ways to transfer the filename (command line being one of them), but all you really need is a process that calls `LoadLibrary`.

**Comment by radare** [comment]
i think we should close this pr and solve the issue in a different way. what your dl thing is doing is loading a lib and calling a symbol, we should be able to do this on other systems. this is injecting a dlopen payload on linux and mac, because remotedylib doesnt exist there. so my proposal is this:

* add dl command to dlopen a library in child process and load their symbols as flags
* add dxc command to execute code in child by doing a call with arguments

this way we will be able to do `dx call 0x8048400` but like `dxs` that allows to pass arguments filling the args in the stack. 

the dl command can be done by adding a new callback on the debugger plugin. that does this magic thing of calling dlopen in the target process. this will probably require to use r_egg to cook those payloads easily. and eventually we may want a `dx` subcommand that accepts egg-lang

**Comment by XVilka** [comment]
@radare Injection of such a payload (and all infractructure about this) - implementing it properly on different system can take a lot of time and certainly wont fit this GSoC then. So may be only as a long term goal for the project... Current approach is much easier, since you can run lib in *nix system anyway.

**Comment by Maijin** [comment]
Yep definitely agree with @XVilka on this^ Way more easier approach currently and works with more cases + it's using rarun2 style approach.

image

**Comment by xarkes** [comment]
I agree with @XVilka and @Maijin, and I think that using rarun2 runlib feature is the easiest way to do it

**Comment by radare** [comment]
But those commands would be nice to have

**Comment by radare** [comment]
so.. ok to close and fill an issue with this info?

**Comment by XVilka** [comment]
But why close? Implementing this in a system-independent manner can take a lot of time. It's better to merge working solution rather than wait for months (years?) for a perfect one. It's always possible to refactor it later.

**Comment by Maijin** [comment]
Yeah but I think we need resolve conflicts

**Comment by radare** [comment]
why? if we dont want this PR merged …

File: libr/core/file.c [hunk 4]
```diff
@@ -784,6 +784,14 @@ R_API RCoreFile *r_core_file_open(RCore *r, const char *file, int flags, ut64 lo
 		flags = 4 | 2;
 	}
 	r->io->bits = r->assembler->bits; // TODO: we need an api for this
+#if __WINDOWS__
+	// XXX Very very dirty
+	// Temporary hack
+	if (strstr (file, ".dll") && !strstr(file, "rundll32.exe") && strstr(file,"dbg://")) {
```
Comment (radare) [inline libr/core/file.c]: missing spaces before (

**Comment by radare** [comment]
ping

**Comment by xarkes** [comment]
ping

**Comment by XVilka** [comment]
@radare @xarkes @Maijin we need to decide this soon, would be nice to have in the release. I should note, that @xarkes wont have time to implement this the way you want.

**Comment by radare** [comment]
Implement what i suggest is the way to go because the rundll thing is something already done in rarun2.

So we want to have a way to load libs in target side, this is already possible with self:// so maybe we should think about allowing to load the r2 preload lib and add a way to communicate the debugger with it, we can do that maybe by allowing r2pipe to talk to the r2 running inside the target process instead of doing all the code injection magic with regg, rbin... enhance dmi, etc

**Comment by XVilka** [comment]
@radare may be, but this is not part of the GSoC. So if you want it just your way - we can close this PR, open a new issue with describing it and wait.

**Comment by radare** [comment]
agree

**Comment by xarkes** [comment]
I'm closing this PR and refering it into this issue https://github.com/radare/radare2/issues/8227

**Pull request closed without merging**